src/crimson/osd/pg_backend.cc

```diff
@@ -930,37 +930,51 @@ PGBackend::get_attr_ierrorator::future<> PGBackend::get_xattrs(
 namespace {
 
 template<typename U, typename V>
-int do_cmp_xattr(int op, const U& lhs, const V& rhs)
+static PGBackend::cmp_xattr_errorator::future<>
+do_cmp_xattr(int op, const U& lhs, const V& rhs)
 {
+  bool result = 0;
   switch (op) {
   case CEPH_OSD_CMPXATTR_OP_EQ:
-    return lhs == rhs;
+    result = (lhs == rhs) ? 1 : 0;
+    break;
   case CEPH_OSD_CMPXATTR_OP_NE:
-    return lhs != rhs;
+    result = (lhs != rhs) ? 1 : 0;
+    break;
   case CEPH_OSD_CMPXATTR_OP_GT:
-    return lhs > rhs;
+    result = (lhs > rhs) ? 1 : 0;
+    break;
   case CEPH_OSD_CMPXATTR_OP_GTE:
-    return lhs >= rhs;
+    result = (lhs >= rhs) ? 1 : 0;
+    break;
   case CEPH_OSD_CMPXATTR_OP_LT:
-    return lhs < rhs;
+    result = (lhs < rhs) ? 1 : 0;
+    break;
   case CEPH_OSD_CMPXATTR_OP_LTE:
-    return lhs <= rhs;
+    result = (lhs <= rhs) ? 1 : 0;
+    break;
   default:
-    return -EINVAL;
+    return crimson::ct_error::invarg::make();
+  }
+  if (!result) {
+    return crimson::ct_error::ecanceled::make();
+  } else {
+    return PGBackend::cmp_xattr_errorator::now();
   }
 }
```

Comment:
Seems like making this return a future is unnecessarily complicated, why not do this in the caller?